docs/in-depth/new-module.md

```diff
@@ -0,0 +1,273 @@
+
+```C++ hl_lines="4 5"
+class HelloWorldModule: public Module {
+public:
+  HelloWorldModule(PoolPtr pool, const ParameterSet& parameters): Module(pool, parameters.getModuleName()) {
+    InputTag inputTag = parameters.get<InputTag>("which_input");
+    input = pool->get<double>(inputTag);
+
+    output = produce<double>("my_output");
+  }
+
+  virtual Status work() override {
+    std::cout << "Hello world!" << std::endl;
+    std::cout << "Our input is: " << *input << std::endl;
+
+    *output = *input * 2;
+
+    return Status::OK;
+  }
+
+private:
+  Value<double> input;
+
+  std::shared_ptr<double> output;
+};
+```
+
+You can now see it's a two-step process to retrieve an input:
+
+  1. We retrieve the parameter `which_input` of type `InputTag`, specified by the user inside the configuration file.
+  2. We use this `InputTag` to retrieve our input.
+
+On the configuration side, the user must now specify a value for the `which_input` parameter:
+
+```lua
+HelloWorldModule.hello = {
+  which_input = "module::output"
+}
+```
+
+!!! tip
+    You can provide a default value when retrieving an argument. This value will be returned if the parameter is not found in the set.
+
+!!! tip
+    You can use the `exists` method of `ParameterSet` to check if a given parameter exists.
+
+### Module registration
+
+When using inputs, outputs or parameters in your modules, extra steps are needed during the registration to inform the system what are your module expectations.
+
+```c++
+REGISTER_MODULE(HelloWorldModule)
+    .Input("which_input")
+    .Output("output");
+```
+
+The above declaration informs the system that your module has one input (`which_input`) and one output (`output`). MoMEMta now expects the user to specify a value for `which_input` when writing the configuration file, and, more specifically, this value must be an `InputTag`.
+
+#### Attributes
+
+Attributes are any parameters which are not inputs. They are used to pass parameters to the module, like for example the current center of mass energy, or the mass of a particle.
+
+You define an attribute when you register the module, by specifying its name and type using the `Attr` or `OptionalAttr` methods, which expect a spec of the form:
+
+```
+<name>: <type-expr> = <default>
+```
+
+where `<name>` begins with a letter and can be composed of alphanumeric characters and underscores, and `<type-expr> `is a type expression of the form [described below](new-module.md#supported-types). An optional `<default>` value can also be specified. If it's the case, the attribute is automatically marked as optional.
+
+Let's imagine our module needs to know what's the mass of the top quark to work. We need to add a new attribute of type `double`:
+
+```c++
+REGISTER_MODULE(HelloWorldModule)
+    .Attr("mass: double")
+    .Input("which_input")
+    .Output("output");
+```
+
+##### Supported types
+
+The following types are supported when registering a new attribute:
```

Comment:
Indeed you're right, but Paths are not intended to be used by others modules than the Looper (it won't work at all) so let's keep that under the hood for the moment 😄